rust/src/lib.rs

```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
+        }));
+        let output_params = Path::new(OsStr::from_bytes(unsafe {
+            slice::from_raw_parts(output_params, output_params_len)
+        }));
+
+        let prover = LocalTxProver::new(spend_params, output_params);
+
+        let pczt_with_proofs = Prover::new(pczt)
+            .create_orchard_proof(&orchard::circuit::ProvingKey::build())
+            .map_err(|e| anyhow!("Failed to create Orchard proof for PCZT: {:?}", e))?
+            .create_sapling_proofs(&prover, &prover)
+            .map_err(|e| anyhow!("Failed to create Sapling proofs for PCZT: {:?}", e))?
+            .finish();
+
+        Ok(FfiBoxedSlice::some(pczt_with_proofs.serialize()))
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Takes a PCZT that has been separately proven and signed, finalizes it, and stores it
+/// in the wallet.
+///
+/// Returns the txid of the completed transaction as a byte array.
+///
+/// # Safety
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must
+///   have an alignment of `1`. Its contents must be a string representing a valid system
+///   path in the operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the
+///   function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_proofs_ptr` must be non-null and valid for reads for `pczt_with_proofs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `pczt_with_proofs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_proofs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_sigs_ptr` must be non-null and valid for reads for `pczt_with_sigs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `pczt_with_sigs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_sigs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_sigs_ptr` must be non-null and valid for reads for `pczt_with_sigs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `pczt_with_sigs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_sigs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `spend_params` must be non-null and valid for reads for `spend_params_len` bytes,
+///   and it must have an alignment of `1`. Its contents must be the Sapling spend proving
+///   parameters.
```

Comment:
> Its contents must be the Sapling spend proving parameters.

I don't think this is a safety requirement.

Reply:
It's also wrong; I copied this documentation from elsewhere but this is a path. I was actually wondering whether this should be passed as a CStr instead.

Reply:
We have other places where we pass a path to parameter files and it should match those, I think.

Reply:
I'm not sure why those other paths aren't passed as C strings.

Reply:
Actually, loading in the parameters from a file is quite expensive iirc. Don't we have a way to preload them?